src/System.Net.Http/src/System/Net/Http/Unix/CurlResponseMessage.cs

```diff
@@ -0,0 +1,247 @@
+            {
+                CheckDisposed();
+                _dataRequestedEvent.Wait(Timeout.InfiniteTimeSpan, _cancellationTokenSource.Token);
+                lock (_lockObject)
+                {
+                    _innerStream = new UnmanagedMemoryStream(pointer, length);
+                    _noMoreDataEvent.Reset();
+                    _dataRequestedEvent.Reset();
+                    _dataReadyEvent.Set();
+                }
+                _noMoreDataEvent.Wait(Timeout.InfiniteTimeSpan, _cancellationTokenSource.Token);;
+            }
+
+            public override bool CanRead
+            {
+                get
+                {
+                    return !_disposed;
+                }
+            }
+
+            public override bool CanSeek
+            {
+                get
+                {
+                    return false;
+                }
+            }
+
+            public override bool CanWrite
+            {
+                get
+                {
+                    return false;
+                }
+            }
+
+            public override long Length
+            {
+                get
+                {
+                    CheckDisposed();
+                    throw new NotSupportedException();
+                }
+            }
+
+            public override long Position
+            {
+                get
+                {
+                    CheckDisposed();
+                    throw new NotSupportedException();
+                }
+
+                set
+                {
+                    CheckDisposed();
+                    throw new NotSupportedException();
+                }
+            }
+
+            public override void Flush()
+            {
+                // Nothing to do.
+            }
+
+            public override int Read(byte[] buffer, int offset, int count)
+            {
+                if (buffer == null)
+                {
+                    throw new ArgumentNullException("buffer");
+                }
+
+                if (offset < 0)
+                {
+                    throw new ArgumentOutOfRangeException("offset");
+                }
+
+                if (count < 0)
+                {
+                    throw new ArgumentOutOfRangeException("count");
+                }
+
+                if ((long) offset + (long) count > (long) buffer.Length)
+                {
+                    throw new ArgumentException("buffer");
+                }
+
+                CheckDisposed();
+
+                do
+                {
+                    lock (_lockObject)
+                    {
+                        // Wait till data is available
+                        if (_innerStream == null)
+                        {
+                            SignalWriter();
+                            continue;
+                        }
+
+                        int bytesRead = _innerStream.Read(buffer, offset, (int)Math.Min((long)count, _innerStream.Length));
```

Comment:
I'm probably being overly cautious here, but consider asserting here that _innerStream is in fact an UnmanagedMemoryStream, with a comment highlighting how disastrous it would be if in the future someone changed _innerStream to be something else where Read could block.